go/pkg/casng/pubsub.go

```diff
@@ -0,0 +1,193 @@
+	t := tag(uuid.New())
+	contextmd.Infof(ctx, log.Level(4), "[casng] pubsub.sub: tag=%s", t)
+
+	// Serialize this block to avoid concurrent map-read-write errors.
+	ps.mu.Lock()
+	subscriber, ok := ps.subs[t]
+	if !ok {
+		subscriber = make(chan any)
+		ps.subs[t] = subscriber
+	}
+	ps.mu.Unlock()
+
+	ps.wg.Add(1)
+	go func() {
+		defer ps.wg.Done()
+		<-ctx.Done()
+
+		contextmd.Infof(ctx, log.Level(4), "[casng] pubsub.unsub: tag=%s", t)
+
+		// Serialize this block to avoid concurrent map-read-write and send-on-closed-channel errors.
+		ps.mu.Lock()
+		delete(ps.subs, t)
+		ps.mu.Unlock()
+
+		close(subscriber)
+	}()
+
+	return t, subscriber
+}
+
+// pub is a blocking call that fans-out a response to all specified (by tag) subscribers sequentially.
+//
+// Returns when all active subscribers have received their copies.
+// Will deadlock if an active subscriber never reveives its copy.
+// Inactive subscribers (expired by cancelling their context) are skipped (their copies are dropped).
+//
+// A busy subscriber does not block others from receiving their copies. It is instead
+// rescheduled for another attempt once all others get a chance to receive.
+// To prevent a temporarily infinite round-robin loop from consuming too much CPU, each subscriber
+// gets at most 10ms to receive before getting rescheduled.
+// Blocking 10ms for every subscriber amortizes much better than blocking 10ms for every
+// iteration on subscribers, even though both have the same worst-case cost.
+// For example, if out of 10 subscribers 5 were busy for 1ms, the attempt will cost ~5ms instead of 10ms.
+func (ps *pubsub) pub(m any, tags ...tag) {
+	_ = ps.pubN(m, len(tags), tags...)
+}
+
+// mpub (multi-publish) delivers the "once" message to a single subscriber then delivers the "rest" message to the rest of the subscribers.
+// It's useful for cases where the message holds shared information that should not be duplicated among subscribers, such as stats.
+func (ps *pubsub) mpub(once any, rest any, tags ...tag) {
+	t := ps.pubOnce(once, tags...)
+	_ = ps.pubN(rest, len(tags)-1, excludeTag(tags, t)...)
+}
+
+// pubOnce is like pub, but delivers the message to a single subscriber.
+// The tag of the subscriber that got the message is returned.
+func (ps *pubsub) pubOnce(m any, tags ...tag) tag {
+	received := ps.pubN(m, 1, tags...)
+	if len(received) == 0 {
+		return ""
+	}
+	return received[0]
+}
+
+// pubN is like pub, but delivers the message to no more than n subscribers. The tags of the subscribers that got the message are returned.
+func (ps *pubsub) pubN(m any, n int, tags ...tag) []tag {
+	if log.V(3) {
+		startTime := time.Now()
+		defer func() {
+			log.Infof("[casng] pubsub.duration: start=%d, end=%d", startTime.UnixNano(), time.Now().UnixNano())
+		}()
+	}
+	if len(tags) == 0 {
+		log.Warning("[casng] pubsub.pub: called without tags")
+		log.V(4).Infof("[casng] pubsub.pub: called without tags: msg=%v", m)
+	}
```

Comment:
is there a point to continue if there are no tags?

Reply:
Good catch